pytensor/graph/basic.py

```diff
@@ -970,6 +970,134 @@ def applys_between(
+
+    .. code-block::
+
+           (c) - n - o
+        n - (n) -'
+
+    * Disconnected condition not returned
+
+    .. code-block::
+
+        (c) - n - o
+         c
+
+    * Disconnected output is present and returned
+
+    .. code-block::
+
+        (c) - (c) - o
+        (o)
+
+    * Condition on itself adds itself
+
+    .. code-block::
+
+        n - (c) - (o/c)
+
+    """
+    # simple case, no additional conditions
+    blockers: Set[Variable] = set()
+    independent_nodes = list()
+    # blockers have known independent nodes and pre-conditions
+    candidates = list(outputs)
+    if not conditions:  # None or empty
+        # just filter out unique variables
+        for node in candidates:
+            if node not in blockers:
+                independent_nodes.append(node)
+                blockers.add(node)
+        # no more actions are needed
+        return independent_nodes
+    blockers.update(conditions)
+    # enforce O(1) check for node in conditions
+    conditions = blockers.copy()
+    # track conditions that are inside the graph to remove disconnected later
+    # preserve order for return
+    conditions_inside = list()
+
+    while candidates:
+        # on any new candidate
+        node = candidates.pop()
+        # check if the node is independent, never go above blockers
+        # blockers are independent nodes and pre-conditions
+        if node in conditions:
+            # The case where node is in conditions so we check if it depends on others
+            # it should be removed from the blockers to check against the rest
+            dependent = variable_is_in_ancestors(node, blockers - {node})
+            # conditions that are present in the graph (not disconnected)
+            # should be added to conditions_inside
+            conditions_inside.append(node)
+            if dependent:
+                # if the condition is still dependent we need to go above,
+                # the search is not yet finished
+                # the node _has_ to have owner to be dependent
+                # so we do not check it
+                # and populate search to go above
+                candidates.extend(node.owner.inputs)
+        else:
+            # A regular node to check
+            dependent = variable_is_in_ancestors(node, blockers)
+            # all regular nodes fall to blockes
+            # 1. it is dependent - further search irrelevant
+            # 2. it is independent - the search node is inside the closure
+            blockers.add(node)
+            # if we've found an independent node and it is not in blockers so far
+            # it is a new indepenent node not present in conditions
+            if not dependent:
+                # we've found an independent node
+                # do not search beyound
+                independent_nodes.append(node)
+            else:
+                # populate search otherwise
+                candidates.extend(node.owner.inputs)
```

Comment:
What if it's a root node without `owner`?

Reply:
hmm

Reply:
Ah this will literally never happen, if a variable is dependent it has to have an owner

Reply:
Can you add a comment saying that? We both had to think about it :)